### PR TITLE
Recommendation 3.12 multiple modifiers sequence

### LINKED ISSUE #158: Recommendation 3.12 If a declaration has multiple modifiers, always follow the sequence below
Recommended sequence:

 ```kotlin
public / internal / protected / private
expect / actual
final / open / abstract / sealed / const
external
override
lateinit
tailrec
crossinline
vararg
suspend
inner
out
enum / annotation
companion
inline / noinline
reified
infix
operator
data
 ```

### PR DESCRIPTION
**Pull request by kentr0w**
## Actions checklist
* [x] Implemented Rule, added Warnings
* [x] Added tests on checks
* [x] Added tests on fixers
* [x] Updated rules-config.json
* [x] Updated available-rules.md
* [x] closes #158

### CODE REVIEW

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/MultipleModifiersSequence.kt [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+                .filter { it.elementType != WHITE_SPACE }
+                .map { Pair(it, MODIFIER_ORDER.indexOf(it.text)) }
+        val sortModifierListOfPair = modifierListOfPair.sortedBy { it.second }
+        modifierListOfPair.forEachIndexed { index, (modifierNode, _) ->
+            if (modifierNode != sortModifierListOfPair[index].first) {
+                WRONG_MULTIPLE_MODIFIERS_ORDER.warnAndFix(configRules, emitWarn, isFixMode,
+                        "this modifier is not in the right position", modifierNode.startOffset) {
+                    val nodeBefore = modifierNode.treeNext
+                    node.removeChild(modifierNode)
+                    node.addChild(KotlinParser().createNode(sortModifierListOfPair[index].first.text), nodeBefore)
```
[the remark this answers is not in the document]
Comment (kentr0w) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/MultipleModifiersSequence.kt:48]: Yes. I tried it differently but nothing worked

[the remark this answers is not in the document]
Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/MultipleModifiersSequence.kt]: Have you tried adding `modifierNode.clone()` as a new node?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/MultipleModifiersSequence.kt [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        emitWarn = emit
+        isFixMode = autoCorrect
+
+        if (node.elementType == MODIFIER_LIST)
+            checkModifierList(node)
+    }
+
+    private fun checkModifierList(node: ASTNode) {
+        val modifierListOfPair = node.getChildren(null)
+                .filter { it.elementType != WHITE_SPACE }
```
[the remark this answers is not in the document]
Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/MultipleModifiersSequence.kt]: I mean, you could just `getChildren(KtTokens.MODIFIER_KEYWORDS)`

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/MultipleModifiersSequence.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (kentr0w) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/MultipleModifiersSequence.kt]: No, but it's working.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/MultipleModifiersSequence.kt [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    private fun checkModifierList(node: ASTNode) {
+        val modifierListOfPair = node.getChildren(KtTokens.MODIFIER_KEYWORDS).toList()
+                .map { Pair(it, MODIFIER_ORDER.indexOf(it.elementType)) }
+        val sortModifierListOfPair = modifierListOfPair.sortedBy { it.second }.map { it.first }
+        modifierListOfPair.forEachIndexed { index, (modifierNode, _) ->
+            if (modifierNode != sortModifierListOfPair[index]) {
+                WRONG_MULTIPLE_MODIFIERS_ORDER.warnAndFix(configRules, emitWarn, isFixMode,
+                        "${modifierNode.text} modifier is not in the right position", modifierNode.startOffset) {
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/MultipleModifiersSequence.kt:51]: I think message like `public should be on position 1 but is on position 2` will be more useful from user's perspective. Now it's hard to reorder modifiers without running fix or diving into sources.

Comment (kentr0w) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/MultipleModifiersSequence.kt]: Fixed

**Review by petertrr — APPROVED** [review]
LGTM

**Pull request merged**